### DISCUSSION TITLE
Building navigation in my first theme

### DISCUSSION THREAD

**Discussion by joshjryan** [BoidCMS/BoidCMS#30] [Q&A] [answered]
Hello,
Should the 'site_nav' action be available to all themes, or do I need to redefine it in my theme? `($App->get_action('site_nav',...)` is returning null, even though I have 3 active Pages).

**Comment by sysadev** [comment, marked as answer]
Hello @joshjryan,

Actually, the way it works is that themes are responsible for actions they need.
And for `site_nav`, it was included only for extensibility, it doesn't work right away. You need to assign links yourself first.

> Note that a second parameter is required for the `get_action` method also. [See for example](https://github.com/BoidCMS/BoidCMS/blob/2519f20258089285a362928ac3b80e5a40bbc879/themes/nimble/blog.php#L20) .

Something like:
```php
<?php global $App ?>
<!DOCTYPE html>
<html>
  <head>
    <title>Webpage</title>
  </head>
  <body>
    <menu-container>
      <!-- The second parameter should be PHP sprintf() compatible format with two placeholders, the first is the link while the second is the text -->
      <?= $App->get_action( 'site_nav', '<menu-item><a href="%s">%s</a><menu-item>' ) ?>
    </menu-container>
  </body>
</html>
```

Below is an example on how you could add links, create a blank plugin or a theme `functions.php` file to test it.

```php
<?php

global $App;

$App->set_action( 'site_nav', 'my_custom_nav' );

function my_custom_nav( string $format ) {
  global $App;
  $nav = '';

  // Selected pages to show in menu
  $selected = array( 'hello-world' );

  // All posts and pages
  $all_pages = $App->data()[ 'pages' ];

  // Loop through
  foreach ( $all_pages as $slug => $fields ) {

    // Choose only the selected ones
    if ( in_array( $slug, $selected ) ) {

      // The slug is relative, let's make it absolute
      $slug = $App->url( $slug );

      // Now make the replacements
      // $slug => The page permalink
      // $fields[ 'title' ] => The page title
      $nav .= sprintf( $format, $slug, $fields[ 'title' ] );

    }

  }

  // You could even add external links directly if you want
  $nav .= sprintf( $format, 'https://example.com', 'Example text' );

  // Finally, return the generated HTML
  return $nav;
}
```

I am going to create a plugin specifically for creating menu items right from the admin panel (eventually).

Good luck with your new theme 👍

**Comment by sysadev** [comment]
Luckily the `Menu` plugin has been completed. [Check it out](https://github.com/BoidCMS/Menu).
